traits/observers/tests/test_trait_event_notifier.py

```diff
@@ -0,0 +1,511 @@
+
+
+def not_prevent_event(event):
+    """ An implementation of prevent_event that does not prevent
+    any event from being propagated.
+    """
+    return False
+
+
+class DummyObservable:
+    """ Dummy implementation of IObservable for testing purposes.
+    """
+
+    def __init__(self):
+        self.notifiers = []
+
+    def _notifiers(self, force_create):
+        return self.notifiers
+
+    def handler(self, event):
+        pass
+
+
+# Dummy target object that is not garbage collected while the tests are run.
+_DUMMY_TARGET = mock.Mock()
+
+
+def create_notifier(**kwargs):
+    """ Convenient function for creating an instance of TraitEventNotifier
+    for testing purposes.
+    """
+    values = dict(
+        handler=mock.Mock(),
+        target=_DUMMY_TARGET,
+        event_factory=mock.Mock(),
+        prevent_event=not_prevent_event,
+        dispatcher=dispatch_here,
+    )
+    values.update(kwargs)
+    return TraitEventNotifier(**values)
+
+
+class TestTraitEventNotifierCall(unittest.TestCase):
+    """ Test calling an instance of TraitEventNotifier. """
+
+    def setUp(self):
+        push_exception_handler(reraise_exceptions=True)
+        self.addCleanup(pop_exception_handler)
+
+    def tearDown(self):
+        pass
+
+    def test_init_and_call(self):
+
+        handler = mock.Mock()
+
+        def event_factory(*args, **kwargs):
+            return "Event"
+
+        notifier = create_notifier(
+            handler=handler, event_factory=event_factory)
+
+        # when
+        notifier(a=1, b=2)
+
+        # then
+        self.assertEqual(handler.call_count, 1)
+        (args, _), = handler.call_args_list
+        self.assertEqual(args, ("Event", ))
+
+    def test_alternative_dispatcher(self):
+        # Test the dispatch is used
+        events = []
+
+        def dispatcher(handler, event):
+            events.append(event)
+
+        notifier = create_notifier(
+            dispatcher=dispatcher,
+            event_factory=mock.Mock(return_value="Event"),
+        )
+
+        # when
+        notifier(a=1, b=2)
+
+        # then
+        self.assertEqual(events, ["Event"])
+
+    def test_prevent_event_is_used(self):
+        # Test prevent_event can stop an event from being dispatched.
+
+        def prevent_event(event):
+            return True
+
+        handler = mock.Mock()
+        notifier = create_notifier(
+            handler=handler,
+            prevent_event=prevent_event,
+        )
+
+        # when
+        notifier(a=1, b=2)
+
+        # then
+        handler.assert_not_called()
+
+    def test_init_check_handler_is_callable_early(self):
+        # Early sanity check to capture misuse
+
+        not_a_callable = None
+        with self.assertRaises(ValueError) as exception_cm:
+            create_notifier(handler=not_a_callable)
+
+        self.assertEqual(
+            str(exception_cm.exception),
+            "handler must be a callable, got {!r}".format(not_a_callable)
+        )
```

Comment:
Is there a reason why you're not using `assertRaisesRegex` (in other tests too)?

Reply:
Okay, I'll leave it up to you then